### PR TITLE
stackdriver: harden against invalid utf8

### PR DESCRIPTION
**Pull request by kyessenov**
Signed-off-by: Kuat Yessenov <kuat@google.com>

Per my experiments, Envoy hardens host and path already, so we sanitize the remaining headers.

### CODE REVIEW

File: extensions/common/context.cc [hunk 1]
```diff
@@ -520,5 +534,19 @@ bool getAuditPolicy() {
   return shouldAudit;
 }
 
+bool sanitizeBytes(std::string* buf) {
```
Comment (lambdai) [inline extensions/common/context.cc:537]: std::string&

File: extensions/common/context.cc [hunk 2]
```diff
@@ -520,5 +534,19 @@ bool getAuditPolicy() {
...
+  char* start = buf->data();
+  const char* const end = start + buf->length();
+  bool modified = false;
+  while (start < end) {
+    if (flatbuffers::FromUTF8(const_cast<const char**>(&start)) < 0) {
```
Comment (lambdai) [inline extensions/common/context.cc:542]: is it possible that `flatbuffers::FromUTF8` access the bytes beyond the last '\0'?

File: extensions/common/context.cc [hunk 3]
```diff
@@ -422,15 +423,26 @@ std::string_view nodeInfoSchema() {
...
+    sanitizeBytes(&request_info->referer);
+  }
+  if (getValue({"request", "useragent"}, &request_info->user_agent)) {
+    sanitizeBytes(&request_info->user_agent);
+  }
+  if (getValue({"request", "id"}, &request_info->request_id)) {
+    sanitizeBytes(&request_info->request_id);
+  }
   std::string trace_sampled;
   if (getValue({"request", "headers", "x-b3-sampled"}, &trace_sampled) &&
```
Comment (lambdai) [inline extensions/common/context.cc:436]: Should `x-b3-sampled` be sanitized as well?

**Review by lambdai — COMMENTED** [review]
It's not clear to me what fields need sanitation and what do not.

I am wondering if `getValue` should by default execute sanitization and explicitly do getValueUnsafe.

File: extensions/common/context.cc (hunk 1 above)
Comment (kyessenov) [reply extensions/common/context.cc:537]: Using references to mutate is not consistent with the rest of this file.

File: extensions/common/context.cc (hunk 2 above)
Comment (kyessenov) [reply extensions/common/context.cc:542]: No, you can see that flatbuffers has an identical loop https://github.com/google/flatbuffers/blob/16836ff95a7f112f66a8c5994a508e6bfec010c9/src/idl_parser.cpp#L76

File: extensions/common/context.cc (hunk 3 above)
Comment (kyessenov) [reply extensions/common/context.cc:436]: No, since this just compares against "1".

File: extensions/common/context.cc (hunk 1 above)
Comment (lambdai) [reply extensions/common/context.cc:537]: I see

**Comment by kyessenov** [comment]
@lambdai There is no requirement that Envoy provides utf-8 for anything. In fact, latin1 encoding is perfectly functional in Envoy. We just don't want to deal with latin1 in stackdriver.

File: extensions/common/context.cc [hunk 4]
```diff
@@ -520,5 +534,19 @@ bool getAuditPolicy() {
...
 }
 
+bool sanitizeBytes(std::string* buf) {
+  char* start = buf->data();
+  const char* const end = start + buf->length();
+  bool modified = false;
+  while (start < end) {
+    if (flatbuffers::FromUTF8(const_cast<const char**>(&start)) < 0) {
+      *start = ' ';
+      start += 1;
```
Comment (lambdai) [inline extensions/common/context.cc:544]: It seem the start could be advanced if the first byte is legit (e.g. the first byte indicate len = 2  b'11000000').
I *think* your goal is to replace b'1100000' by SPACE and start from the byte after b'11000000'. 

https://github.com/google/flatbuffers/blob/df2df21ec1be2468106fed10c1533eacc1cf0d2e/include/flatbuffers/util.h#L528


my imagined nagative case is a two byte string: b'11000000' b'11000000' 

Can you confirm?

Comment (kyessenov) [reply extensions/common/context.cc:544]: Good point. Missed this with the cast. PTAL again, and updated test case with \C0.

Comment (lambdai) [reply extensions/common/context.cc:545]: LGTM!

**Review by lambdai — APPROVED** [review]
Thank you for the update!

**Comment by bianpengyuan** [comment]
@kyessenov  curious, how does envoy deal invalid utf8 at host or url path?

**Comment by kyessenov** [comment]
@bianpengyuan Envoy issues a codec error and returns 400 Bad Request

**Pull request merged**